synapse/lib/drive.py

```diff
@@ -0,0 +1,536 @@
+        # update the info with the newest version info...
+        if datavers >= curvers:
+            info.update(versinfo)
+            rows.append((LKEY_INFO + bidn, s_msgpack.en(info)))
+
+        self.slab.putmulti(rows, db=self.dbname)
+
+        return info, versinfo
+
+    def getItemData(self, iden, vers=None):
+        '''
+        Return a (versinfo, data) tuple for the given iden. If
+        version is not specified, the current version is returned.
+        '''
+        return self._getItemData(s_common.uhex(iden), vers=vers)
+
+    def _getItemData(self, bidn, vers=None):
+
+        if vers is None:
+            info = self._getItemInfo(bidn)
+            vers = info.get('version')
+
+        versindx = getVersIndx(vers)
+        versbyts = self.slab.get(LKEY_VERS + bidn + versindx, db=self.dbname)
+        if versbyts is None: # pragma: no cover
+            return None
+
+        databyts = self.slab.get(LKEY_DATA + bidn + versindx, db=self.dbname)
+        if databyts is None: # pragma: no cover
+            return None
+
+        return s_msgpack.un(versbyts), s_msgpack.un(databyts)
+
+    def delItemData(self, iden, vers=None):
+        return self._delItemData(s_common.uhex(iden), vers=vers)
+
+    def _delItemData(self, bidn, vers=None):
+
+        info = self._reqItemInfo(bidn)
+        if vers is None:
+            vers = info.get('version')
+
+        versindx = getVersIndx(vers)
+
+        self.slab.delete(LKEY_VERS + bidn + versindx, db=self.dbname)
+        self.slab.delete(LKEY_DATA + bidn + versindx, db=self.dbname)
+
+        # back down or revert to 0.0.0
+        if vers == info.get('version'):
+            versinfo = self._getLastDataVers(bidn)
+            if versinfo is None:
+                info['size'] = 0
+                info['version'] = (0, 0, 0)
+                info.pop('updated', None)
+                info.pop('updater', None)
+            else:
+                info.update(versinfo)
+
+        self.slab.put(LKEY_INFO + bidn, s_msgpack.en(info), db=self.dbname)
+        return info
+
+    def _getLastDataVers(self, bidn):
+        for lkey, byts in self.slab.scanByPrefBack(LKEY_VERS + bidn, db=self.dbname):
+            return s_msgpack.un(byts)
+
+    async def getItemDataVersions(self, iden):
+        '''
+        Yield data version info in reverse created order.
+        '''
+        bidn = s_common.uhex(iden)
+        pref = LKEY_VERS + bidn
+        for lkey, byts in self.slab.scanByPrefBack(pref, db=self.dbname):
+            yield s_msgpack.un(byts)
+            await asyncio.sleep(0)
+
+    def getTypeSchema(self, typename):
+        byts = self.slab.get(LKEY_TYPE + typename.encode(), db=self.dbname)
+        if byts is not None:
+            return s_msgpack.un(byts)
+
+    async def setTypeSchema(self, typename, schema, callback=None):
+
+        vtor = s_config.getJsValidator(schema)
+
+        self.validators[typename] = vtor
+
+        lkey = LKEY_TYPE + typename.encode()
```

Comment:
Should probably check that typename length is < 510 to avoid the lmdb keysize error